### PR TITLE
fix(project): genomic projection of c.* variants in the 3'UTR poly-A region (NM_003002.2; #790 residual)

### LINKED ISSUE #797: fix(project): genomic projection of c.* variants in the 3'UTR poly-A region (NM_003002.2; #790 residual)
## Problem

A `c.*` variant whose 3'UTR position falls in a transcript's **poly-A tail** cannot be projected to the genome — the poly-A tail is post-transcriptional and has no genomic coordinate, so the exon→genome map (correctly) does not cover it.

Surfaced by #790 (PR #795): the NM_003002.2 derived exon structure covers the genomic core (poly-A excluded), so:

```
NG_012337.1(NM_003002.2):c.*824del      expected="NG_012337.1:g.13948del"  err="project: Invalid coordinates: Cannot map tx position 1364 to genome"
NG_012337.1(NM_003002.2):c.*824+10del   expected="NG_012337.1:g.13958del"  err="project: Invalid coordinates: Cannot map tx position 1364 to genome"
```

Both remain in `baseline-failures/genomic.txt`.

## What was tried (and why it's not trivial)

Extending the last exon into contiguous downstream genome (so the poly-A region gets coordinates) was attempted during #790 and **reverted** — it produced a *wrong* genome coordinate (`g.13965del` for both, vs mutalyzer's distinct `g.13948`/`g.13958`), because the sibling-anchored 3'UTR differs from the target's. mutalyzer maps these positions by a base-agnostic coordinate walk into contiguous genome using the *target's own* 3'UTR structure, which a sibling anchor doesn't reproduce.

## Direction (decide in design)

- Map a `c.*` poly-A-region position to contiguous downstream genome using the **target transcript's own** 3'UTR genomic extent (not a sibling's), when derivable; otherwise decline cleanly rather than emit a wrong coordinate. 
- This is narrow (the only corpus rows today are the two SDHD `c.*824` rows) but is the residual #790 explicitly carved out.

Refs #790, #795, #487/#672.

### PR DESCRIPTION
**Pull request by nh13**
Closes #797.

## Summary

A `c.*` 3'UTR variant whose transcript position lands in a transcript's post-transcriptional **poly-A tail** could not project to the genome. The #790-derived exon→genome structure (correctly) strips the poly-A tail (it has no genomic coordinate), so such a position has no containing exon and the CDS-aware mapper declined with `Cannot map tx position N to genome`. The two SDHD corpus rows were stuck in `baseline-failures/genomic.txt`:

```
NG_012337.1(NM_003002.2):c.*824del      expected NG_012337.1:g.13948del
NG_012337.1(NM_003002.2):c.*824+10del   expected NG_012337.1:g.13958del  (normalizes to c.*834del)
```

## Decision: correct-map (not decline)

mutalyzer maps these by a **base-agnostic coordinate walk** into contiguous downstream genome using **the transcript's own 3'-terminal exon anchor** — not the sibling structure used only to *anchor* the #790 derivation. A naive last-exon extension was tried and reverted during #790 because it produced a single wrong coordinate (`g.13965`) for both rows; the correct contiguous walk yields the distinct `g.13948` / `g.13958`.

The walk is derivable from the target's own structure: the derived last exon's genomic bounds were transferred by aligning NM_003002.2's own mRNA (poly-A stripped, genome-readback validated, `mismatch_fraction = 0.0`); the sibling NM_003002.3 only anchored the alignment. Extending contiguously from that anchor reproduces mutalyzer's coordinates exactly (verified against the prepared reference: per-position substitution probe gives `g = 13124 + c.*N`, so `c.*824 → g.13948`, `c.*834 → g.13958`).

## Implementation

- `CdotTranscript::tx_to_genome_extending_polya` — geometry-only contiguous walk past the 3'-terminal exon (plus: `genome_end + delta`; minus: `genome_start - 1 - delta`), declining on a pre-terminal hole or a minus-strand underflow below the 1-based genome origin. In-exon positions delegate to the unchanged `tx_to_genome`.
- `VariantProjector::try_extend_polya_to_genome` — invoked only when the CDS-aware mapper declines a `c.*` position; folds an optional positive `c.*N+k` offset linearly (matching the normalizer's `c.*N+k → c.*(N+k)`) and **bounds the walk to the transcript's TRUE FASTA length** (the derived exon map's length excludes the stripped tail), so a position past the real poly-A tail declines cleanly rather than walking into non-transcript genome. An over-extension past the parent's placed span still self-declines in `reanchor_genome_to_parent`.
- `MockProvider::get_sequence_length` now mirrors `get_sequence`'s routing (contig dispatch, then `get_transcript` with unversioned fallback) so it serves transcript lengths, matching the production `MultiFastaProvider` contract.

## Non-regression

The change can only turn a prior mapper decline into a successful map — it never alters an existing result. Verified on the genomic conformance axis (backup manifest): it removes **exactly** the two SDHD rows (56 → 54 fails) and introduces **zero** new failures. Normal CDS / 5'UTR / intronic / genomic-core 3'UTR projection is unchanged. The two `c.*824` lines are removed from `baseline-failures/genomic.txt`; the rows carry no genomic-axis annotation and their expected `genomic` field already equals the now-correct output, so they tally as plain passes.

## Testing

- Unit (`src/data/cdot.rs`): plus/minus poly-A walk, minus underflow decline, pre-terminal-hole decline.
- Projector (`src/project/projector.rs`): plus & minus poly-A walk, `c.*N+k` offset folding (both strands), genomic-core 3'UTR unchanged, beyond-tail decline.
- `cargo nextest run --features dev` — green (7530 tests). Manifest-gated genomic axis with `FERRO_MANIFEST=<prepared reference>` confirms `c.*824del → g.13948del` and `c.*824+10del → g.13958del`.
- `cargo clippy --features dev --all-targets -- -D warnings` — clean.

Refs #790, #795.

## Summary by CodeRabbit

* **New Features**
  * Extended transcript-to-genome mapping to support coordinates beyond the last exon into the 3′ poly-A tail.
  * Added poly-A-aware projection fallback for `c.*` endpoints, including correct multi-axis handling and UTR behavior.

* **Bug Fixes**
  * Improved poly-A handling so `c.*` endpoints that previously declined now project downstream safely; non-terminal/unsupported cases still fail as before.
  * Enhanced mock reference sequence-length lookup to better resolve genomic contigs vs transcripts.

* **Tests**
  * Added unit tests for plus/minus poly-A extension, underflow/overflow safety, and gap/decline edge cases.
  * Updated normalization baseline-failure fixtures.

### CODE REVIEW

**Comment by nh13** [comment]
@coderabbitai resume

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

File: src/project/projector.rs [hunk 1]
```diff
@@ -994,6 +994,79 @@ impl<P: ReferenceProvider + Clone> VariantProjector<P> {
...
+        }
+        // Fold the (downstream) offset into the effective tx coordinate.
+        let tx_pos = base_tx_pos.checked_add(offset)?;
+        // Gate to the transcript's TRUE length (FASTA) so the walk stays within
+        // the real poly-A tail. Decline if the length is unavailable.
+        let true_tx_length = self.provider.get_sequence_length(transcript_id).ok()?;
+        if tx_pos >= true_tx_length {
+            return None;
+        }
+        cdot_tx.tx_to_genome_extending_polya(tx_pos)
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs]: Fixed. `try_extend_polya_to_genome` now validates the walked coordinate against the contig bounds: it declines if the result is `0` (a 1-based genome coordinate is never zero) or exceeds the contig length looked up via `get_sequence_length(&cdot_tx.contig)`. The transcript-length guard above is a transcript-frame bound; this adds the contig-frame bound the plus-strand `genome_end + delta` walk could otherwise overrun. The contig-length lookup is treated leniently (decline only on a definite overflow, not on an unknown length) so a provider that cannot serve a contig length does not regress the real SDHD case.

File: src/project/projector.rs [hunk 2]
```diff
@@ -1237,7 +1310,31 @@ impl<P: ReferenceProvider + Clone> VariantProjector<P> {
...
+                            // propagate the original decline otherwise — so beyond
+                            // the tail we never walk into non-transcript genome.
+                            if let Some(g) =
+                                self.try_extend_polya_to_genome(cdot_tx, &transcript_id, &p)
+                            {
+                                return Ok(g);
+                            }
+                            return Err(e);
+                        }
+                    };
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs:1562]: Good catch — this was a genuine gap in the multi-axis path. Confirmed and fixed.

The poly-A genome coordinate is outside the cdot exon span by construction, so when a `c.*` poly-A input was routed through the full `project`/`project_variant`/`project_variant_all` APIs, `project_single_inner` re-derived the coding axis from that synthetic coordinate via genome to CDS and the span check rejected it as `TranscriptNotOverlapping` (verified with a probe test before the fix).

Fix: `project_single_inner` now detects a transcript-coordinate input whose genome pivot lands entirely outside `[tx_genome_start, tx_genome_end)` and short-circuits to a new `polya_multiaxis_projection` helper that keeps the input's own coding form and reports the re-anchored walked coordinate as the genomic axis (a `c.*` 3'UTR/poly-A position has no protein/RNA/noncoding consequence to derive from a position with no genomic exon context, so those axes are `None`). A genuine off-transcript g. input is unaffected — the short-circuit only triggers for transcript-coordinate inputs, so a real non-overlapping g. variant still raises `TranscriptNotOverlapping`.

Added regressions through the public `project_variant` API on both strands: `project_variant_polya_region_multiaxis_short_circuits` and `project_variant_minus_polya_region_multiaxis_short_circuits`.

File: src/project/projector.rs [hunk 3]
```diff
@@ -6970,6 +7067,327 @@ mod tests {
...
+                exon_cigars: Vec::new(),
+                cached_introns: OnceLock::new(),
+                protein_id: Some("NP_POLYA.1".to_string()),
+            });
+            // 999 leading Ns so HGVS g.1000 == 0-based index 999; the downstream
+            // genome past the exon (g.1012, g.1013, …) is real genomic sequence
+            // (the poly-A walk is base-agnostic, so the bases there don't matter).
+            let prefix = "N".repeat(999);
+            let suffix = "N".repeat(100);
+            provider.add_genomic_sequence("chr1", format!("{}TTTATGCGCGTAAA{}", prefix, suffix));
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs]: Fixed. Both poly-A fixtures now register the genomic backbone under the `NC_000001.11` output accession as well as `chr1` (the cdot contig), so the genomic-frame renormalization `project_to_genomic` runs on its output can actually fetch bases instead of silently falling back to the un-normalized form. The flanks are also now non-repeating: a shared `polya_genomic_sequence()` helper builds an `ACGT`-cycle prefix, the 12-base exon core, and a `CAGT`-cycle downstream tail (no homopolymer at any single position), so a single-base del at the asserted poly-A coordinates (g.1012 / g.1013 plus, g.999 / g.998 minus) stays put under 3'-shifting — normalization is genuinely exercised without moving the asserted coordinate. Applied to both `make_polya_test_provider_and_projector` and `make_minus_polya_test_provider_and_projector`.

File: src/reference/mock.rs [hunk 4]
```diff
@@ -607,10 +607,33 @@ impl ReferenceProvider for MockProvider {
...
+            }
+        }
+        // 2. Otherwise treat it as a transcript accession, via `get_transcript`
+        //    so the unversioned-fallback / version-substitution paths apply
+        //    exactly as in `get_sequence`.
+        if let Ok(tx) = self.get_transcript(id) {
+            if let Some(seq) = &tx.sequence {
+                return Ok(seq.len() as u64);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/reference/mock.rs:635]: Fixed. `get_sequence_length` now returns `ReferenceNotFound` immediately from the known-contig branch when no genomic sequence is loaded, instead of falling through to the transcript lookup — matching `get_sequence`, which returns from the contig branch immediately. This keeps strict contig routing so a known contig id with no loaded genomic sequence can no longer resolve to an ambiguous transcript length.

**Comment by nh13** [comment]
@coderabbitai resume

**Comment by nh13** [comment]
@coderabbitai review

File: src/data/cdot.rs [hunk 5]
```diff
@@ -803,6 +803,63 @@ impl CdotTranscript {
...
+        let last = self.exons.last()?;
+        let (genome_start, genome_end, tx_end) = (last[0], last[1], last[3]);
+        if tx_pos < tx_end {
+            // Falls before the terminal exon's end but had no exon — a gap/hole,
+            // not the poly-A region. Decline rather than guess.
+            return None;
+        }
+        let delta = tx_pos - tx_end;
+        match self.strand {
+            Strand::Plus => Some(genome_end + delta),
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/data/cdot.rs]: Fixed. `tx_to_genome_extending_polya`'s plus-strand arm now uses `genome_end.checked_add(delta)`, so a poly-A walk that would overflow `u64` declines (`None`) instead of wrapping/panicking before the caller's contig-bound check sees the coordinate. Added a unit regression `test_tx_to_genome_extending_polya_plus_overflow_declines` pinning a 3'-terminal exon at the top of the `u64` range: `tx_end` returns `Some(u64::MAX)` and one base further declines.

File: src/project/projector.rs [hunk 6]
```diff
@@ -2823,6 +2935,30 @@ impl<P: ReferenceProvider + Clone> VariantProjector<P> {
...
+                if let Some(projection) = self.polya_multiaxis_projection(
+                    normalized,
+                    transcript_id,
+                    &projected_genome,
+                    gene_symbol.clone(),
+                ) {
+                    return Ok(projection);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs:3216]: Fixed. The poly-A multi-axis short-circuit is now gated on `matches!(normalized, HgvsVariant::Cds(_))` (was the over-broad `!matches!(normalized, HgvsVariant::Genome(_))`), so only a genuine `c.*` endpoint can take the bypass. `n.`/`r.` inputs fall through to the normal overlap path and still raise the canonical `TranscriptNotOverlapping` decline. The `polya_multiaxis_projection` helper was tightened to match: its `coding` builder now accepts only the `Cds` arm (the dead `Tx`/`Rna` arms removed), so the gate cannot be silently re-loosened in future. Added a public-API regression `project_variant_offexon_noncoding_input_is_not_polya_short_circuited` proving an off-exon `n.` input declines cleanly rather than being smuggled into the `.coding` slot.

**Comment by nh13** [comment]
Addressing the outside-diff finding "Thread the poly-A fallback through fan-out seeding" (`src/project/projector.rs` ~310-326).

Fixed. The `HgvsVariant::Cds(c)` arm of `extract_contig_and_pos` (the seed-position extractor for `project_variant_all` fan-out) previously hard-failed on a `c.*` poly-A endpoint, because `cds_to_genome_on_build` declines for a position in the stripped post-transcriptional poly-A tail — so `project_variant_all` propagated the error and never enumerated the transcript. It now falls back to `try_extend_polya_to_genome` (which self-gates on a genuine `c.*` poly-A endpoint); when that confirms a poly-A endpoint, the stab query is seeded from an **in-transcript** anchor — the 3'-terminal exon's genomic start, which is inside the exon span on either strand — rather than the walked downstream coordinate (which is past the exon end by construction and would land the stab off the transcript). The per-transcript projection then takes the multi-axis short-circuit. When the position is not a poly-A endpoint the original decline is propagated unchanged.

This covers both routing concerns raised:
- **NC-parented `c.*`** poly-A inputs now seed fan-out correctly (previously hit the raw mapper and failed).
- **NG/LRG-parented `c.*`** poly-A inputs arrive as `HgvsVariant::Cds` with a `genomic_context` parent (not a de-anchored `g.`), so they go through this same `Cds` arm — `deanchor_genomic_parent_input` only rewrites genomic (`g.`) inputs, so a `c.*`-on-NG input is never de-anchored to a synthetic outside-span `g.` coordinate.

Added the required `project_variant_all` edge-case regression `project_variant_all_polya_region_enumerates_transcript`: a `c.*4del` poly-A input now enumerates `NM_POLYA.1` through the public `project_variant_all` API and reports `g.1012del` (genomic axis) with the `c.*4del` coding axis preserved — it would error pre-fix.

Verification: `cargo nextest run --features dev` → 7577 passed; `cargo clippy --features dev --all-targets -- -D warnings` clean. The manifest-backed genomic conformance axis (`c.*824`/SDHD poly-A) is unchanged — divergence lists are byte-identical before vs after these changes (0 new failures).

**Comment by nh13** [comment]
@coderabbitai resume

**Comment by nh13** [comment]
@coderabbitai review

File: src/project/projector.rs [hunk 7]
```diff
@@ -1237,7 +1351,31 @@ impl<P: ReferenceProvider + Clone> VariantProjector<P> {
...
+                            // exon. `try_extend_polya_to_genome` self-gates (a `c.*`
+                            // position whose base lands in the poly-A region, an
+                            // optional positive downstream offset, and an effective
+                            // tx position within the transcript's TRUE length, i.e.
+                            // inside the real poly-A tail) and returns None to
+                            // propagate the original decline otherwise — so beyond
+                            // the tail we never walk into non-transcript genome.
+                            if let Some(g) =
+                                self.try_extend_polya_to_genome(cdot_tx, &transcript_id, &p)
+                            {
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs]: Fixed in d5f87c0. The `map_pos` closure in `project_to_genomic_nc` is shared across `Cds`/`Tx`/`Rna` inputs (all reshaped to `CdsPos` with `utr3` carrying their own downstream/3'UTR flag), so the poly-A walk fallback was reachable for `n.`/`r.`. Captured `is_cds_input = matches!(variant, HgvsVariant::Cds(_))` before the closure and gated `try_extend_polya_to_genome` behind it. Added two `project_to_genomic` regressions proving off-exon `n.*` and `r.*` endpoints decline (`TranscriptNotOverlapping`/`InvalidCoordinates`) rather than emitting a genomic coordinate via the poly-A path.

File: src/project/projector.rs [hunk 8]
```diff
@@ -2823,6 +2961,33 @@ impl<P: ReferenceProvider + Clone> VariantProjector<P> {
...
+        // Only a `c.*` (CDS) input whose pivot lands entirely past the span
+        // qualifies: the poly-A tail fallback is keyed on a `c.*` 3'UTR endpoint
+        // (`try_extend_polya_to_genome` self-gates on `p.utr3`), so it has no
+        // meaning for `n.`/`r.` inputs. A genuine off-transcript `g.` input — or a
+        // non-overlapping `n.`/`r.` input — must still raise
+        // `TranscriptNotOverlapping` via the normal path below.
+        if matches!(normalized, HgvsVariant::Cds(_)) {
+            let outside = |p: &GenomePos| p.base < tx_genome_start || p.base >= tx_genome_end;
+            if outside(&g_start) && outside(&g_end) {
+                if let Some(projection) = self.polya_multiaxis_projection(
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs:3207]: Fixed in d5f87c0. Confirmed reachable: on the fixture, `c.*3` maps inside the stripped terminal exon (g.1011) and `c.*4` via the poly-A walk (g.1012), so `outside(&g_start) && outside(&g_end)` was false and `map_position` rejected the single outside (poly-A) endpoint as `TranscriptNotOverlapping`. Changed the gate to `outside(&g_start) || outside(&g_end)`. This is safe in this `Cds`-only branch: `projected_genome` only holds an off-span coordinate that `project_to_genomic_nc` produced via the (now `c.`-gated) poly-A walk — `cds_to_genome` otherwise yields an in-exon coordinate — so an outside endpoint here is necessarily a validated poly-A endpoint, and `polya_multiaxis_projection` re-anchors that already-walked interval rather than re-deriving from it. Added plus- and minus-strand boundary-spanning regressions (`c.*3_*4delinsGG`; delins avoids homopolymer repeat-collapse that would otherwise shift the interval and obscure the boundary shape). Both fail under the old `&&` gate and pass under `||`.

**Comment by nh13** [comment]
Re: outside-diff comment on `src/project/projector.rs` 1587-1602 ("Do not de-anchor NG/LRG poly-A `c.*` inputs to synthetic genome seeds") — fixed in d5f87c0.

`deanchor_genomic_parent_input`'s `c.` branch called `project_to_genomic_nc`, which now succeeds for a poly-A `c.*` via the contiguous walk and produced an OFF-exon `NC_` coordinate; de-anchoring to that synthetic `Genome` seeded the `project_variant_all` stab off the transcript span, so enumeration found nothing and the locus was mis-projected.

Fix: added `cds_input_has_polya_endpoint(variant)` (resolves the transcript on the input's build and checks whether either endpoint passes `try_extend_polya_to_genome`). In the NG_/LRG_ `c.` de-anchor branch, when a confirmed poly-A endpoint is detected we now keep the transcript-coordinate (`c.`) form and just signal the parent for re-framing instead of de-anchoring to the synthetic `Genome`. The `c.` fan-out seed path (`extract_contig_and_pos`) already anchors poly-A endpoints to the 3'-terminal exon, so enumeration + the multi-axis short-circuit proceed correctly — mirroring how a bare-`NM_` poly-A `c.*` already fans out.

Regression: `project_variant_all_ng_parented_polya_enumerates_under_parent` projects `NG_POLYA.1(NM_POLYA.1):c.*4del` through `project_variant_all`, asserting `NM_POLYA.1` is enumerated with the coding axis framed under the `NG_` parent. It fails without the fix (transcript not enumerated) and passes with it. (The conformance `axis_genomic` cases `NG_012337.1(NM_003002.2):c.*824del` / `c.*824+10del` go through the single-transcript `project_to_genomic` re-anchor path and remain passing, unchanged by this fix.)

**Comment by nh13** [comment]
@coderabbitai resume

**Comment by nh13** [comment]
@coderabbitai review

File: src/project/projector.rs [hunk 9]
```diff
@@ -6970,6 +7283,855 @@ mod tests {
...
+            // all instead of raising TranscriptNotOverlapping.
+            let g = proj
+                .genomic
+                .as_ref()
+                .expect("boundary-spanning poly-A projection must carry a genomic axis")
+                .to_string();
+            assert!(
+                g.starts_with("NC_000001.11"),
+                "expected an NC_-frame genomic axis, got: {g}"
+            );
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/project/projector.rs:8002]: Fixed in 92e54a6. Both boundary-spanning tests now pin the exact normalized genomic interval and edit, not just the accession prefix:

- `project_variant_polya_boundary_spanning_short_circuits` (plus): asserts `:g.1011_1012delinsGG` — start g.1011 inside the stripped terminal exon, end g.1012 the first poly-A walk coordinate, literal `GG` insertion.
- `project_variant_minus_polya_boundary_spanning_short_circuits` (minus): asserts `:g.999_1000delinsCC` — `c.*3`→g.1000 (inside exon, higher coord on minus), `c.*4`→g.999 (poly-A walk, lower coord), with `GG` reverse-complemented to `CC`.

Coordinates were captured from the actual public-path output and confirmed against the fixtures' geometry, so a wrong endpoint or a missing reverse-complement now fails the test. Test-only change; full poly-A suite green, clippy clean.

**Comment by nh13** [comment]
@coderabbitai resume

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
Fixed in 6f726e1. The genomic baseline ledger is now regenerated from the authoritative machine output (`/tmp/ferro-xfail/genomic.txt` via the manifest-gated `axis_genomic` run) and sorted to match the committed convention. The stale `NG_012337.1(NM_012459.2):c.-1_*1del` entry is gone — it was never in the machine-generated FAIL set. Net delta vs `main` is now exactly the two SDHD `c.*824` poly-A removals (the stated #797 conformance delta), with zero new failures (verified: genomic axis 54 FAILs = base − 2; all five other mutalyzer axes unchanged).

**Comment by nh13** [comment]
@coderabbitai resume

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review
